AMDGPU/SILowerI1Copies process phi incomings in specific order

petar-avramovic opened the pull request:
When merging lane masks value from block that is always visited first
(PrevReg in buildMergeLaneMasks) needs to exist. For PrevReg to exist,
basic block that should contain PrevReg definition must be processed
first.
If block A dominates block B, block A must be processed first.
This can be achieved by sorting phi incomings.

Sorting of phi incomings makes no changes for phis created by SDAG
because SDAG adds phi incomings as it selects basic blocks in reversed
post order traversal.

This change is required by upcoming lane mask merging implementation for
GlobalISel that leaves phi incomings as they are in IR.

petar-avramovic committed:
AMDGPU/SILowerI1Copies: refactor phi incoming handling in lowerPhis [NFC]
Incoming block, incoming Register and updated incoming Register that
correspond to the same incoming of a phi are kept on same index in
different vectors.
Use structure with fields: block, register and updated register instead.

nhaehnle reviewed (commented):
Thanks! Surely the test can be simplified though? Apart from that and the inline comment, this LGTM

jayfoad commented on llvm/lib/Target/AMDGPU/SILowerI1Copies.cpp:
```diff
@@ -581,10 +587,24 @@ bool SILowerI1Copies::lowerPhis() {
 
-      IncomingBlocks.push_back(IncomingMBB);
-      IncomingRegs.push_back(IncomingReg);
+      Incomings.push_back({IncomingReg, IncomingMBB, Register{}});
```
Maybe emplace_back instead?

jayfoad commented on llvm/lib/Target/AMDGPU/SILowerI1Copies.cpp:
```diff
@@ -581,10 +587,24 @@ bool SILowerI1Copies::lowerPhis() {
+    // dominates() can't be used since it is not strict weak ordering so we use
+    // DFSNumIn, DFS visitation order for nodes in the dominator tree, from
+    // MachineDominatorTree.
+    DT->getBase().updateDFSNumbers();
```
Do you need to call this every time around the Vreg1Phis loop?

petar-avramovic replied:
Thanks, this can be moved before the loop.

petar-avramovic commented on llvm/lib/Target/AMDGPU/SILowerI1Copies.cpp:
```diff
@@ -581,10 +587,24 @@ bool SILowerI1Copies::lowerPhis() {
+      Incomings.push_back({IncomingReg, IncomingMBB, Register{}});
     }
 
+    // When building merge-LaneMasks, if block A dominates block B, block A
+    // must be processed first. If there is no dominate relation order does
+    // not matter since there will be no lane mask merging.
+    // To ensure this we need to sort incomings by some criteria
+    // dominates() can't be used since it is not strict weak ordering so we use
+    // DFSNumIn, DFS visitation order for nodes in the dominator tree, from
+    // MachineDominatorTree.
```
Yes, you are right. I would like to look into this again, global-isel gets register 0 as PrevReg in some cases. I will update the patch soon.

petar-avramovic committed:
AMDGPU/SILowerI1Copies process phi incomings in specific order

When merging lane masks, value from block that is always visited first
(PrevReg in buildMergeLaneMasks) needs to exist because we do on-the-fly
constant folding. For PrevReg to exist, basic block that should contain
PrevReg definition must be processed first. Sort the incomings such that
incoming values that dominate other incoming values are processed first.

Sorting of phi incomings makes no changes for phis created by SDAG
because SDAG adds phi incomings as it selects basic blocks in reversed
post order traversal.

This change is required by upcoming lane mask merging implementation
for GlobalISel that leaves phi incomings as they are in IR.

arsenm commented:
> Just a quick question what should I do if I want to merge two separate patches, open two PRs?

Yes. There's 3rd party tooling to help with this but I'm not sure there's a consensus on the best way to deal with this yet

petar-avramovic commented:
I reopened two PRs with intention to have patches merged one by one
patch2 https://github.com/llvm/llvm-project/pull/72375
patch1 https://github.com/llvm/llvm-project/pull/72374

petar-avramovic commented:
Patches were merged via https://github.com/llvm/llvm-project/pull/72374 and https://github.com/llvm/llvm-project/pull/72375
